### ISSUE TITLE
About performance

### ISSUE THREAD

**Issue by bityangke** [GuYuc/WS-DAN.PyTorch#1]
Hi
Can this implementation reproduce the performance in the original paper?
Thanks!

**Comment by GuYuc** [comment]
I have not run the code on dataset shown in this paper (e.g. CUB-200, FGVC-Aircraft), instead, I've got comparative results on some FGVC 6 datasets, and it outperforms other methods such as DFL-CNN and RA-CNN on the same datasets.

**Comment by morning4346** [comment]
I have run the code on standfordDog dataset,but it only achieve 82%,but the peformance in this paper have
reached 92.1%,can you tell me why?

**Comment by GuYuc** [comment]
@liyinhua4346 Hi, thank you for taking experiments on these classical datasets by using these code. This is not an official implementation, and maybe I didn't totally understand the actual way of training, e.g. the training order of 3 kinds of data (raw images, cropped images, dropped images), the usage of feature centers, etc. The authors promised to publicize their code in their paper, so let's wait for their code for details.

**Comment by zhengyu19921215** [comment]
InceptionV3 does not participate in training, so the  accuary is low？？

**Comment by GuYuc** [comment]
@zhengyu19921215 Hi, i believe all params of the network were updated during training, so did you encounter problems that make inception_v3 stop updating in training process? Please contact me if these unexpected problems really occur when running the code.

**Comment by zhengyu19921215** [comment]
I mean cropped images, dropped image is not involved in training, and the total loss should be calculated in the calculation of losses, rather than a single loss backward

**Comment by seobbro** [comment]
I think the paper missing something. I achieve 81.4% accuracy in CUB-200 dataset using this code.

**Comment by cahya-wirawan** [comment]
Hi, I also can't reproduce the performance on "196 Stanford Cars" dataset. The paper stated it achieved 94.5% accuracy, but my best accuracy is only 89.0% using this implementation with Inceptionv3 (and original train:test distribution of 8144:8041).

**Comment by xyy19920105** [comment]
in CUB-200 dataset, I only can get about 84% accuracy

**Comment by erfaneshrati** [comment]
Anyone got any success in reproducing the CUB-200 accuracy?

**Comment by tau-yihouxiang** [comment]
There should be sign-sqrt normalization and l2 normalization after obtaining a feature matrix, by referring Bilinear Pooling. Also, I multiply the normalized feature by 100x to accelerate the training process. The source code will be released this month.

**Comment by libing125** [comment]
I also can't reproduce the result on Stanford Dogs. I only achieved 80%.

**Comment by GuYuc** [comment]
@tau-yihouxiang Thank you. I should have noticed the concept of Bilinear Pooling mentioned in the paper and paid more attention on classification modules instead of merely adding a simple fully connected layer.

**Comment by WonderSeven** [comment]
I can't reproduce the result of CUB, I only achieve 80%, need I magnitude the hyperparameters by myself?

**Comment by GuYuc** [comment]
@WonderSeven Hi. You may need to add a sign-sqrt normalization and l2 normalization after obtaining a feature matrix, as @tau-yihouxiang had mentioned in this issue. I am too busy to review my code and do experiments by now, but I will update the code and experiments in the future.

**Comment by ashiqimran** [comment]
How did you split dataset on train, val and test? The original has only train and test for CUB_200_2010.

**Comment by Lucas1347** [comment]
@tau-yihouxiang Hi, can you explain what is sign-sqrt normalization? I can not find the definition of sign-sqrt normalization. Thx~

**Comment by ashiqimran** [comment]
Does anyone here succeed to replicate the results of original paper?

**Comment by WonderSeven** [comment]
@GuYuc Quiet thanks for your release and suggestions, I have two doubts: 1) the training procedure, as you mentationed above, the order is right?  Or maybe this three loss could be conbined together rather than in the one-by-one way like now.  2) the work of attention regularization, i don't konw whether this could learn a fixed part of body (e.g. head) even with large posture difference. Am I missing some thing?

**Comment by WonderSeven** [comment]
@GuYuc Quiet thanks for your release and suggestions, I have two doubts: 1) three training losses. I combine this three losses by addition and achieved (81.7) which is greater than raw(80.*).  2) the work of attention regularization, I don't konw whether this could learn a fixed part of body (e.g. head) even with large posture difference. Am I missing some thing?

**Comment by Marcus1996-dot** [comment]
Hi,thanks for the implementation. I have  run the code on dataset CUB-200 and only get get about 84% accuracy.and then I  add a sign-sqrt normalization and l2 normalization in Bilinear Attention Pooling after obtaining a feature matrix.but it occur a bug in the code .sorry i can't ressolve it.could you please help me?
![火柴截图20190809090521237](https://user-images.githubusercontent.com/53923029/62747003-f592d180-ba84-11e9-9ce1-5806d8ea46b7.jpg)

**Comment by gen0924** [comment]
@GuYuc thanks for your code , but, i can only achieved 81.5% on CUB-200, coudle you show me how to apply the sign-sqrt normalization and l2 normalization ? Thanks

**Comment by amireshratifar-zz** [comment]
The author promised to publish the code in July but it didn't happen.

**Comment by ashiqimran** [comment]
@Marcus1996-dot, Hi,could you share the training, validation and test sizes that you use to get 84%? I didn't get that much accuracy yet. Thanks in advance.

**Comment by tau-yihouxiang** [comment]
Hi, all. The official code has been released in https://github.com/tau-yihouxiang/WS_DAN. Thank you all for discussion.

**Comment by Marcus1996-dot** [comment]
@ashiqimran .Hi,i think that the validation process in train_wsdan.py means the test，so i set the training size  and validation size to 5994 and 5794 respectively in a rewrite dataset.py.if i mistake this part,please inform me.

**Comment by ashiqimran** [comment]
@Marcus1996-dot, Yes, what you said makes sense to me and I also did that. But I got less than 80% on bird dataset. Any luck of implementing sign-sqrt normalization?

**Comment by Marcus1996-dot** [comment]
@tau-yihouxiang ,@ashiqimran ,actually,I added a sign-sqrt normalization and l2 normalization after obtaining a feature matrix as follows.This time i achieved 87% on CUB-200，but i still can't reproduce the result in the paper.Do you think i mistake this part or i still miss something? Thanks in advance.
![image](https://user-images.githubusercontent.com/53923029/62998413-0df85700-be9e-11e9-99dd-05323d93393d.png)

**Comment by amireshratifar-zz** [comment]
@Marcus1996-dot it would be great if you could share ur repo so that we also be able to reproduce 87%.

**Comment by Marcus1996-dot** [comment]
@amireshratifar ,actually,i just added a sign-sqrt normalization and l2 normalization in the wsdan.py,i think you could just do the same in  the code here.it is really not that different.

**Comment by xyy19920105** [comment]
@Marcus1996-dot ,I did as you said, but I only achieved 83% on CUB-200 in batch size 12.

**Comment by narrowsnap** [comment]
@xyy19920105 @amireshratifar  I achieved 87.73, and the code is at https://github.com/narrowsnap/ws_dan_jupyter. But I find the center loss doesn't work. Hope help for you.

**Comment by xyy19920105** [comment]
@narrowsnap so great! Thanks

**Comment by albert0713** [comment]
After epoch 80, Val Acc is (0.52, 2.71). Is it normal?
